network/adaptive.go

```diff
@@ -0,0 +1,172 @@
+package network
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/ethereum/go-ethereum/rlp"
+)
+
+// CapabilityId defines a unique type of capability
+// @acud concrete enough for ya?
+type CapabilityId uint64
+
+// Capability contains a bit vector of flags that define what capability a node has in a specific module
+// The module is defined by the Id.
+type Capability struct {
+	Id  CapabilityId
+	Cap []bool
+}
+
+// NewCapability initializes a new Capability with the given id and specified number of bits in the vector
+func NewCapability(id CapabilityId, bitCount int) *Capability {
+	return &Capability{
+		Id:  id,
+		Cap: make([]bool, bitCount),
+	}
+}
+
+// Set switches the bit at the specified index on
+func (c *Capability) Set(idx int) error {
+	l := len(c.Cap)
+	if idx > l-1 {
+		return fmt.Errorf("index %d out of bounds (len=%d)", idx, l)
+	}
+	c.Cap[idx] = true
+	return nil
+}
+
+// Unset switches the bit at the specified index off
+func (c *Capability) Unset(idx int) error {
+	l := len(c.Cap)
+	if idx > l-1 {
+		return fmt.Errorf("index %d out of bounds (len=%d)", idx, l)
+	}
+	c.Cap[idx] = false
+	return nil
+}
+
+// String implements Stringer interface
+func (c Capability) String() (s string) {
+	s = fmt.Sprintf("%d:", c.Id)
+	for _, b := range c.Cap {
+		if b {
+			s += "1"
+		} else {
+			s += "0"
+		}
+	}
+	return s
+}
+
+// IsSameAs returns true if the given Capability object has the identical bit settings as the receiver
+func (c *Capability) IsSameAs(cp *Capability) bool {
+	if cp == nil {
+		return false
+	}
+	return isSameBools(c.Cap, cp.Cap)
+}
+
+func isSameBools(left []bool, right []bool) bool {
+	if len(left) != len(right) {
+		return false
+	}
+	for i, b := range left {
+		if b != right[i] {
+			return false
+		}
+	}
+	return true
+}
+
+// Capabilities is the collection of capabilities for a Swarm node
+// It is used both to store the capabilities in the node, and
+// to communicate the node capabilities to its peers
+type Capabilities struct {
+	idx  map[CapabilityId]int
+	Caps []*Capability
+	mu   sync.Mutex
+}
+
+// NewCapabilities initializes a new Capabilities object
+func NewCapabilities() *Capabilities {
+	return &Capabilities{
+		idx: make(map[CapabilityId]int),
+	}
+}
+
+// adds a capability to the Capabilities collection
+func (c *Capabilities) add(cp *Capability) error {
+	if _, ok := c.idx[cp.Id]; ok {
+		return fmt.Errorf("Capability id %d already registered", cp.Id)
+	}
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.Caps = append(c.Caps, cp)
+	c.idx[cp.Id] = len(c.Caps) - 1
+	return nil
+}
+
+// gets the capability with the specified module id
+// returns nil if the id doesn't exist
+func (c Capabilities) get(id CapabilityId) *Capability {
```

Comment:
consider changing the return signature to `(*Capability, bool)`? not a blocker though, but i think is more clear to handle in the calling context

Reply:
As the code stands now there doesn't seem to be need for this. How about we change it if it appears unclear somewhere.

Reply:
https://travis-ci.org/ethersphere/swarm/jobs/571902858#L491